pkg/operator/hive/hive.go

```diff
@@ -17,87 +17,87 @@ limitations under the License.
-	// Parse yaml for all Hive objects:
+
 	asset := assets.MustAsset("config/manager/deployment.yaml")
 	hLog.Debug("reading deployment")
 	hiveDeployment := resourceread.ReadDeploymentV1OrDie(asset)
 
-	// Set owner refs on all objects in the deployment so deleting the operator CRD
-	// will clean everything up:
-	// NOTE: we do not cleanup the CRDs themselves so as not to destroy data.
-
-	if err := controllerutil.SetControllerReference(instance, hiveDeployment, r.scheme); err != nil {
-		hLog.WithError(err).Info("error setting owner ref")
-		return err
-	}
-
-	expectedDeploymentGen := int64(0)
-	currentDeployment := &appsv1.Deployment{}
-	foundCurrentDeployment := false
-	err := r.Get(context.Background(), types.NamespacedName{Name: hiveDeployment.Name, Namespace: hiveDeployment.Namespace}, currentDeployment)
-	if err != nil && !errors.IsNotFound(err) {
-		hLog.WithError(err).Error("error looking up current deployment")
-		return err
-	} else if err == nil {
-		expectedDeploymentGen = currentDeployment.ObjectMeta.Generation
-		foundCurrentDeployment = true
-	}
-
 	if r.hiveImage != "" {
 		hiveDeployment.Spec.Template.Spec.Containers[0].Image = r.hiveImage
 		// NOTE: overwriting all environment vars here, there are no others at the time of
 		// writing:
-		hiveDeployment.Spec.Template.Spec.Containers[0].Env = []corev1.EnvVar{
-			{
-				Name:  images.HiveImageEnvVar,
-				Value: r.hiveImage,
-			},
+		hiveImageEnvVar := corev1.EnvVar{
+			Name:  images.HiveImageEnvVar,
+			Value: r.hiveImage,
 		}
+
+		hiveDeployment.Spec.Template.Spec.Containers[0].Env = append(hiveDeployment.Spec.Template.Spec.Containers[0].Env, hiveImageEnvVar)
 	}
 
-	// ApplyDeployment does not check much of the Spec for changes. Do some manual
-	// checking and if we see something we care about has changed, force an update
-	// by changing the expected deployment generation.
-	if foundCurrentDeployment && currentDeployment.Spec.Template.Spec.Containers[0].Image !=
-		hiveDeployment.Spec.Template.Spec.Containers[0].Image {
-		hLog.WithFields(log.Fields{
-			"current": currentDeployment.Spec.Template.Spec.Containers[0].Image,
-			"new":     hiveDeployment.Spec.Template.Spec.Containers[0].Image,
-		}).Info("overriding deployment image")
-		expectedDeploymentGen = expectedDeploymentGen - 1
+	// Ensure the hive admin roles are configured properly:
+	clientConfig := util.GenerateClientConfigFromRESTConfig("anything", r.restConfig)
+	kubeconfig, err := clientcmd.Write(*clientConfig)
+	if err != nil {
+		hLog.WithError(err).Error("error serializing kubeconfig")
+		return err
 	}
+	h := resource.NewHelper(kubeconfig, hLog)
 
-	var changed bool
-	_, changed, err = resourceapply.ApplyDeployment(r.kubeClient.AppsV1(),
-		recorder, hiveDeployment, expectedDeploymentGen, false)
+	s := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme,
+		scheme.Scheme)
+
+	// TODO: it would be nice to be able to log if there were changes or not
```

Comment:
It looks like this is doable using the ToPrinter function passed to the command. I'll look into implementing and return an "action taken" (created, changed, unchanged) along with the error.